examples/include/crypto-accelerator.p4

```diff
@@ -0,0 +1,136 @@
+/*
+Copyright 2022 Advanced Micro Devices, Inc
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/// Crypto accelerator Extern
+enum bit<8> crypto_algorithm_e {
+    AES_GCM = 1
+}
+enum bit<8> crypto_results_e {
+    SUCCESS = 0,
+    AUTH_FAILURE = 1,
+    HW_ERROR = 2
+}
+
+enum bit<8> crypto_error_action_e {
+    NO_ACTION = 0,
+    DROP_PACKET = 1,
+    SEND_TO_PORT = 2
+}
+
+#define ICV_AFTER_PAYLOAD ((int<32>)-1)
+
+/// The crypto_accelerator engine used in this example uses AES-GCM algorithm.
+/// It is assumed to be agnostic to wire protocols i.e. does not understand protocol 
+/// specific headers like ESP, AH etc
+///
+/// Note that crypto accelerator does not modify the packet outside the payload area and ICV
+///     Any wire-protocol header, trailer add/remove is handled by P4 pipeline
```

Comment:
Today there is no standard way for a P4 pipeline to add/remove trailers at the end of long pipelines.  The best workaround for short packets is to completely parse every byte of the short packet, and add/remove headers at the end, but that is not a good solution for packets that can be MTU in size.

Does your solution propose new capabilities in PNA and/or P4 to add/remove trailers?

Reply:
Alternately, one could avoid the need for handling trailers in P4 pipeline completely by bundling together trailer add/remove into the encrypt/decrypt block logic.

Reply:
The architecture can do anything it wants with trailers, that's how ethernet trailers actually work.
But you could add some externs to manipulate them. Symmetric with packet_in, but for trailers.

Reply:
There are multiple ways to add/manipulate trailers. This capability is useful in endpoints (PNA where protocols like ESP, RDMA do use trailers). One method can be - payload is defined as a variable length header. Parser can extract it (based on length computed from ipHeader.totalLen or any other means available). Deparser similarly can add it back to packet and then add trailer headers after it. This will not need any new support in P4 language.
Many targets and/or target compilers allow large extractions as long as control functions do not access the extracted bytes as in this case.

Reply:
For targets that need trailer processing by the accelerator, more methods can be added to this extern object for handling the trailers.. but those will be target specific, so I have not shown them here.

Reply:
Accessing/using trailers is done by more than one protocols so it is possible to define target-specific externs to add/remove/modify trailers that are generically useful outside IPSec ESP. Since the inline accelerators are after deparser, it is expected that trailers are added *before* packet reaches the accelerator. For crypto accelerators, trailers may be just payload that will be encrypted.
From the division of work perspective, I did not want to include this capability in the accelerator. In case it is required for specific targets, we can define target-specific methods for it as mentioned in earlier comment.